model-client/src/commonMain/kotlin/org/modelix/model/client2/ModelClientV2.kt

```diff
@@ -491,32 +492,55 @@ private fun URLBuilder.appendPathSegmentsEncodingSlash(vararg components: String
 fun VersionDelta.getAllObjects(): Map<String, String> = objectsMap + objects.associateBy { HashUtil.sha256(it) }
 
 suspend fun HttpResponse.readVersionDelta(): VersionDeltaStream {
-    return if (contentType()?.match(VersionDeltaStream.CONTENT_TYPE) == true) {
-        val content = bodyAsChannel()
-        val versionHash = checkNotNull(content.readUTF8Line()) { "No objects received" }
-        val versionObject = content.readUTF8Line()
-        return if (versionObject == null) {
-            VersionDeltaStream(versionHash, emptyFlow())
-        } else {
-            VersionDeltaStream(
-                versionHash,
-                flow {
-                    emit(versionHash to versionObject)
-                    while (true) {
-                        val key = content.readUTF8Line() ?: break
-                        val value = checkNotNull(content.readUTF8Line()) { "Object missing for hash $key" }
-                        emit(key to value)
-                    }
-                },
-            )
-        }
+    return if (contentType()?.match(VersionDeltaStreamV2.CONTENT_TYPE) == true) {
+        return readVersionDeltaStreamV2()
+    } else if (contentType()?.match(VersionDeltaStream.CONTENT_TYPE) == true) {
+        return readVersionDeltaStreamV1()
     } else {
-        body<VersionDelta>().asStream()
+        val versionDelta = body<VersionDelta>()
+        versionDelta.objectsMap.forEach {
+            HashUtil.checkObjectHash(it.key, it.value)
+        }
+        versionDelta.asStream()
     }
 }
 
 fun HttpRequestBuilder.useVersionStreamFormat() {
-    headers.set(HttpHeaders.Accept, VersionDeltaStream.CONTENT_TYPE.toString())
+    headers[HttpHeaders.Accept] = VersionDeltaStreamV2.CONTENT_TYPE.toString()
+    // Add CONTENT_TYPE_VERSION_DELTA_V1 so that newer clients cant talk with older servers.
+    headers.append(HttpHeaders.Accept, VersionDeltaStream.CONTENT_TYPE.toString())
+}
+
+private suspend fun HttpResponse.readVersionDeltaStreamV1(): VersionDeltaStream {
+    val content = bodyAsChannel()
+    val versionHash = checkNotNull(content.readUTF8Line()) { "No objects received" }
+    val versionObject = content.readUTF8Line()
+    return if (versionObject == null) {
+        VersionDeltaStream(versionHash, emptyFlow())
+    } else {
+        VersionDeltaStream(
+            versionHash,
+            flow {
+                emit(versionHash to versionObject)
+                while (true) {
+                    val key = content.readUTF8Line() ?: break
+                    val value = checkNotNull(content.readUTF8Line()) { "Object missing for hash $key" }
+                    HashUtil.checkObjectHash(key, value)
+                    emit(key to value)
+                }
+            },
+        )
+    }
+}
+
+private suspend fun HttpResponse.readVersionDeltaStreamV2(): VersionDeltaStream {
+    val content = bodyAsChannel()
+    return VersionDeltaStreamV2.decodeVersionDeltaStreamV2(content).toVersionDeltaStream()
```

Comment:
Why do we have a different implementation pattern for the format here than for V1?

Reply:
Because I wanted to implement encoding and decoding independant of the `ModelServer`/`ModelClient` to make it reusabel and testabel in isolation.
See `VersionDeltaStreamV2`